crates/shared/src/price_estimation/quasimodo.rs

```diff
@@ -0,0 +1,347 @@
+
+        let mut tokens = self.base_tokens.tokens().clone();
+        tokens.insert(query.sell_token);
+        tokens.insert(query.buy_token);
+        tokens.insert(self.native_token);
+        let tokens: Vec<_> = tokens.drain().collect();
+
+        let token_infos = self.token_info.get_token_infos(&tokens).await;
+
+        let tokens = tokens
+            .iter()
+            .map(|token| {
+                let info = token_infos.get(token).cloned().unwrap_or_default();
+                (
+                    *token,
+                    TokenInfoModel {
+                        decimals: info.decimals,
+                        alias: info.symbol,
+                        normalize_priority: Some(if self.native_token == query.buy_token {
+                            1
+                        } else {
+                            0
+                        }),
+                        ..Default::default()
+                    },
+                )
+            })
+            .collect();
+
+        let (sell_amount, buy_amount) = match query.kind {
+            OrderKind::Buy => (U256::max_value(), query.in_amount),
+            OrderKind::Sell => (query.in_amount, U256::one()),
+        };
+
+        let orders = BTreeMap::from([(
+            0,
+            OrderModel {
+                sell_token: query.sell_token,
+                buy_token: query.buy_token,
+                sell_amount,
+                buy_amount,
+                allow_partial_fill: false,
+                is_sell_order: query.kind == OrderKind::Sell,
+                fee: FeeModel {
+                    amount: U256::from(GAS_PER_ORDER) * gas_cost,
+                    token: self.native_token,
+                },
+                cost: CostModel {
+                    amount: U256::from(GAS_PER_ORDER) * gas_cost,
+                    token: self.native_token,
+                },
+                is_liquidity_order: false,
+                mandatory: true,
+            },
+        )]);
+
+        let token_pair = TokenPair::new(query.sell_token, query.buy_token).unwrap();
+        let pairs = self.base_tokens.relevant_pairs([token_pair].into_iter());
+
+        let amms = self
+            .pools
+            .fetch(pairs, Block::Recent)
+            .await?
+            .iter()
+            .map(|pool| AmmModel {
+                parameters: AmmParameters::ConstantProduct(ConstantProductPoolParameters {
+                    reserves: BTreeMap::from([
+                        (pool.tokens.get().0, pool.reserves.0.into()),
+                        (pool.tokens.get().1, pool.reserves.1.into()),
+                    ]),
+                }),
+                fee: BigRational::from((
+                    BigInt::from(*pool.fee.numer()),
+                    BigInt::from(*pool.fee.denom()),
+                )),
+                cost: CostModel {
+                    amount: U256::from(GAS_PER_UNISWAP) * gas_cost,
```

Comment:
They're not in this PR. This is because we have different representations for pools in solver and orderbook. I think the plan is to unify representations, and then add support for all pools in this estimator.